sys/kern/src/task.rs

```diff
@@ -820,47 +820,55 @@ pub fn check_task_id_against_table(
 /// Selects a new task to run after `previous`. Tries to be fair, kind of.
 ///
 /// If no tasks are runnable, the kernel panics.
-pub fn select(previous: usize, tasks: &[Task]) -> usize {
-    priority_scan(previous, tasks, |t| t.is_runnable())
-        .expect("no tasks runnable")
+pub fn select(previous: usize, tasks: &[Task]) -> &Task {
+    match priority_scan(previous, tasks, |t| t.is_runnable()) {
+        Some((_index, task)) => task,
+        None => panic!(),
+    }
 }
 
+/// Scans the task table to find a prioritized candidate.
+///
 /// Scans `tasks` for the next task, after `previous`, that satisfies `pred`. If
 /// more than one task satisfies `pred`, returns the most important one. If
 /// multiple tasks with the same priority satisfy `pred`, prefers the first one
-/// in order after `previous`, mod `tasks.len()`.
+/// in order after `previous`, mod `tasks.len()`. Finally, if no tasks satisfy
+/// `pred`, returns `None`
 ///
 /// Whew.
 ///
 /// This is generally the right way to search a task table, and is used to
 /// implement (among other bits) the scheduler.
 ///
-/// # Panics
-///
-/// If `previous` is not a valid index in `tasks`.
+/// On success, the return value is the task's index in the task table, and a
+/// direct reference to the task.
 pub fn priority_scan(
     previous: usize,
     tasks: &[Task],
     pred: impl Fn(&Task) -> bool,
-) -> Option<usize> {
-    uassert!(previous < tasks.len());
-    let search_order = (previous + 1..tasks.len()).chain(0..previous + 1);
-    let mut choice = None;
-    for i in search_order {
-        if !pred(&tasks[i]) {
+) -> Option<(usize, &Task)> {
+    let mut pos = previous;
+    let mut choice: Option<(usize, &Task)> = None;
+    for _step_no in 0..tasks.len() {
+        pos = pos.wrapping_add(1);
+        if pos >= tasks.len() {
+            pos = 0;
+        }
+        let t = &tasks[pos];
```

Comment:
similarly, there's probably a few instructions and a panic site that could be shaved off here by using `get_unchecked`, given that the `if` immediately above this line ensures that `pos` is always in bounds. On the other hand, LLVM might be smart enough to figure that out here --- the code that ensures it's in bounds is straightforward enough that I wouldn't be shocked if the compiler could figure it out...

Reply:
Yeah, this (and basically all other cases where an access is dominated by a naive `i < slice.len()`) is optimized out. I'm trying to only apply unsafe in cases where I can't figure out how to get the panics removed from the safe code, and this code specifically does not produce a panic rn.

Reply:
Ah, in that case, let's leave it as is! Good job, LLVM!